File: Scripts/2024/restToRestMotion.m
function [time, position, velocity, acceleration] = restToRestMotion(totalTime, accelerationValue, speedValue, accelerationTime)
    % This function simulates the motion profile of a robot going from
    % Rest-to-Rest
    %
    % Parameters:
    % - totalTime: Total time of the motion profile.
    % - accelerationValue: Acceleration magnitude during the acceleration phase.
    % - speedValue: Constant speed during the coast phase.
    % - accelerationTime: Time duration of the acceleration phase.
    %
    % Outputs:
    %  1. Plot of position wrt time
    %  2. Plot of speed wrt time
    %  3. Plot of acceleration wrt time
    
    % Calculate coast time
    coastTime = totalTime - 2 * accelerationTime;

    % Time vector
    t1 = linspace(0, accelerationTime, 100);
    t2 = linspace(accelerationTime, accelerationTime + coastTime, 100);
    t3 = linspace(accelerationTime + coastTime, totalTime, 100);
    time = [t1, t2, t3];

    % Acceleration profile
    a1 = ones(1, 100) * accelerationValue;
    a2 = zeros(1, 100);
    a3 = -ones(1, 100) * accelerationValue;
    acceleration = [a1, a2, a3];

    % Velocity profile
    v1 = cumtrapz(t1, a1);
    v2 = ones(1, 100) * speedValue;
    v3 = cumtrapz(t3, a3) + v2(end);
    velocity = [v1, v2, v3];

    % Position profile (updated integration)
    x1 = cumtrapz(t1, v1);
    x2 = cumtrapz(t2, v2) + x1(end);
    x3 = cumtrapz(t3, v3) + x2(end);
    position = [x1, x2, x3];

    % Plotting the profiles
    figure;
    subplot(3, 1, 1);
    plot(time, position, 'b.-');
    title('Position vs Time');
    xlabel('Time (s)');
    ylabel('Position');

    subplot(3, 1, 2);
    plot(time, velocity, 'g.-');
    title('Velocity vs Time');
    xlabel('Time (s)');
    ylabel('Velocity');

    subplot(3, 1, 3);
    plot(time, acceleration, 'r.-');
    title('Acceleration vs Time');
    xlabel('Time (s)');
    ylabel('Acceleration');
end

File: Scripts/2025/python_scripts/robobox/rest_to_rest_motion.py
import numpy as np
import matplotlib.pyplot as plt

def rest_to_rest_motion(total_time: float, acceleration_value: float, speed_value: float, acceleration_time: float):
    """
    Displays the motion profile of a robot rest-to-rest motion

    @param total_time: The total time of the motion profile (in s).
    @param acceleration_value: The acceleration magnitude during the acceleration phase (in distance_unit/s^2).
    @param speed_value: The constant speed during the coast phase (in units of distance_unit/s).
    @param acceleration_time: The time duration of the acceleration phase (in s).
    """
    # Coast time
    coast_time = total_time - 2*acceleration_time
    
    # Time vector
    t1 = np.linspace(0, acceleration_time, 100)
    t2 = np.linspace(acceleration_time, acceleration_time + coast_time, 100)
    t3 = np.linspace(acceleration_time + coast_time, total_time, 100)
    time = np.concatenate([t1, t2, t3])
    
    # Acceleration profile
    a1 = np.ones_like(t1) * acceleration_value
    a2 = np.zeros_like(t2)
    a3 = np.ones_like(t3) * -acceleration_value
    acceleration = np.concatenate([a1, a2, a3])
    
    # Velocity profile
    v1 = np.cumsum(a1)*(t1[1] - t1[0])  
    v2 = np.ones_like(t2)*speed_value
    v3 = np.cumsum(a3)*(t3[1] - t3[0]) + v2[-1]
    velocity = np.concatenate([v1, v2, v3])
    
    # Position profile
    x1 = np.cumsum(v1) * (t1[1] - t1[0]) 
    x2 = np.cumsum(v2) * (t2[1] - t2[0]) + x1[-1]
    x3 = np.cumsum(v3) * (t3[1] - t3[0]) + x2[-1]
    position = np.concatenate([x1, x2, x3])

    # Display
    fig, axis = plt.subplots(3, 1, figsize=(8, 8))

    axis[0].plot(time, position, 'r-')
    axis[0].set_title('Position profile')
    axis[0].set_xlabel('Time (s)')
    axis[0].set_ylabel('Position')
    axis[0].grid()

    # Plot velocity vs time
    axis[1].plot(time, velocity, 'g-')
    axis[1].set_title('Velocity profile')
    axis[1].set_xlabel('Time (s)')
    axis[1].set_ylabel('Velocity')
    axis[1].grid()

    # Plot acceleration vs time
    axis[2].plot(time, acceleration, 'b-')
    axis[2].set_title('Acceleration profile')
    axis[2].set_xlabel('Time (s)')
    axis[2].set_ylabel('Acceleration')
    axis[2].grid()
    
    plt.tight_layout()
    plt.show()


